### PR TITLE
Series.map Performance Improvement When Using Dictionaries

### LINKED ISSUE #46248: PERF: pd.Series.map slow for huge dictionary
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

This issue is a duplicate of [#34717](https://github.com/pandas-dev/pandas/issues/34717) since I do not believe it has been resolved. Using the same example as the previous issue, I can reproduce the poor performance of `pd.Series.map` with large dictionaries, as opposed to using `pd.Series.apply` with `dict.get`.

```py

In [3]: pd.__version__
Out[3]: '1.4.1'

In [4]: %cpaste
Pasting code; enter '--' alone on the line to stop or use Ctrl-D.
:n = 1000000
domain = np.arange(0, n)
ranges = domain+10
maptable = pd.Series(ranges, index=domain).sort_index().to_dict()
query_vals = pd.Series([1,2,3])::::
:--

In [5]: %timeit query_vals.map(maptable)
343 ms ± 1.61 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit query_vals.apply(lambda x: maptable.get(x))
81.4 µs ± 862 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

At the very least I think it would be useful to include some disclaimer in the documentation for `pd.Series.map` explaining this drawback more clearly and presenting the alternative solution, but of course a fix for this would be ideal. I would be happy to take this issue if I can be pointed in the right direction.

### Installed Versions

<details>

```
INSTALLED VERSIONS
------------------
commit           : 06d230151e6f18fdb8139d09abf539867a8cd481
python           : 3.9.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.13.0-28-generic
Version          : #31~20.04.1-Ubuntu SMP Wed Jan 19 14:08:10 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.1
numpy            : 1.21.2
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 58.0.4
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.31.1
pandas_datareader: None
bs4              : None
bottleneck       : 1.3.2
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : 2.8.1
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
```

</details>

### Prior Performance

_No response_

### PR DESCRIPTION
**Pull request by ryansdowning**
- [x] closes #46248
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Hello, first time contributing to pandas so if I did something wrong, please let me know!

In the discussion for #46248 I proposed a very simple solution using `collections.defaultdict`. However, after implementing the change, I found a performance improvement of 10x, but that was still a far stretch from the 1000-10000x improvement I was expecting. After more testing I found that the majority of the time was now being spent by copying the dictionary to the `defaultdict`. To avoid unnecessarily copying the dictionary's data, I use `dict.get` with a default of `np.nan` for dictionaries that do not implement the `__missing__` method ~~added `ReadOnlyNanDefaultDict` which gives us the behavior of `defaultdict(lambda: np.nan, dictionary_object)` without copying the data. As given by the name, it is also read only since that is all that is needed by `IndexOpsMixin._map_values`.~~  With this change, I see the huge performance boost of 1000-10000x.

This is a slight deviation from the `defaultdict` solution proposed in the issue, but it performs much better and behaves in the same way.

Lastly, I tried to add type hints for `IndexOpsMixin._map_values` but could not find the proper way of typing `Series` without causing a circular import. Is `"Series"` the way to go, or is there something better?

### CODE REVIEW

File: pandas/core/base.py [hunk 1]
```diff
@@ -184,6 +183,24 @@ class SpecificationError(Exception):
     pass
 
 
+class ReadOnlyNanDefaultDict:
```
Comment (jreback) [inline pandas/core/base.py:186]: move this to where it is used. (e.g. inside the function itself).

File: pandas/core/base.py [hunk 2]
```diff
@@ -824,20 +841,17 @@ def _map_values(self, mapper, na_action=None):
...
+                # Dictionary does not have a default. Since the default
+                # behavior for missing values in a dictionary mapping is
+                # to map them to np.nan, it is safe to convert the
+                # dictionary to a defaultdict which returns np.nan for
+                # missing keys for efficiency (GH #46248). For even better
+                # performance, ReadOnlyNanDefaultDict provides the
+                # necessary functionality, without copying the dictionary.
+                dict_with_default = ReadOnlyNanDefaultDict(mapper)
+
+            mapper = lambda x: dict_with_default[x]
```
Comment (jreback) [inline pandas/core/base.py:854]: can we just use the Series path?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by ryansdowning** [comment]
I am running into a behavior inconsistency that is leading to some tests failing due to this change. 

```py
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: from collections import defaultdict

In [4]: s = pd.Series([1, 2, np.nan])

In [5]: mapper = {1: "a", 2: "b", np.nan: "c"}

In [6]: default_mapper = defaultdict(str, mapper)

In [7]: s.map(mapper)
Out[7]: 
0    a
1    b
2    c
dtype: object

In [8]: s.map(default_mapper)
Out[8]: 
0    a
1    b
2     
dtype: object
``` 
On the most recent main branch, mapping for dictionaries vs dictionaries with missing methods behave differently with `np.nan` as keys. I don't know enough about the inner workings of the map vectorization, but I suspect it has something to do with `lib.map_infer` which I cannot locate the source for. Nevertheless, I am unsure how to precede from here given this inconsistent behavior.

**Comment by rhshadrach** [comment]
@ryansdowning - not sure if this might help, but one has to be very careful with `np.nan` in dictionaries. E.g.

```
map = {np.nan: 1}
arr = np.asarray([np.nan])
print(map.get(arr[0]))
```

produces `None`. I'd guess that the internals are treating `dict` differently than other Mappings. The code for `lib.map_infer` is in `pandas._libs.lib.pyx` starting (roughly) on L2866.

**Comment by mroeschke** [comment]
Thanks for the effort on this PR so far, but it appears to have gone stale. If interested in continuing, please merge in main, address the failures in the CI, and we can reopen.

**Pull request closed without merging**